modules/pipeline/providers/reconciler/reconcile.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reconciler
+
+import (
+	"context"
+	"runtime/debug"
+	"sync"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/modules/pipeline/providers/reconciler/rutil"
+	"github.com/erda-project/erda/modules/pipeline/providers/reconciler/schedulabletask"
+	"github.com/erda-project/erda/modules/pipeline/spec"
+)
+
+func (r *provider) ReconcileOnePipeline(ctx context.Context, pipelineID uint64) {
+	// make pipelineID-level context to support cancel reconcile
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	defer func() {
+		if err := recover(); err != nil {
+			r.Log.Errorf("panic while reconcile one pipeline until done, cancel reconcile, pipelineID: %d, err: %v", pipelineID, err)
+			debug.PrintStack()
+		}
+	}()
```

Comment:
I think cancel do first is better? if panic

Reply:
It makes sense.
defer recover at the beginning of the file.
defer cancel direct after declaration.